Tril tests for ibits2f, etc.

knn-k opened the pull request:
This commit implements Tril test cases for ibits2f, etc.

Signed-off-by: knn-k <konno@jp.ibm.com>

knn-k commented:
The `UsingConst` variants of `fbits2i` and `dbits2l` are disabled as they fail.
`std::snprintf(inputTrees, 256, "(fconst %f)", value)` generates `(fconst 0.000000)` when the `value` is 1.17549e-38, and that leads to unexpected results.

Is there any way to handle cases like this?

knn-k commented on fvtest/compilertriltest/TypeConversionTest.cpp:
```diff
@@ -670,3 +672,274 @@ INSTANTIATE_TEST_CASE_P(TypeConversionTest, Int64ToInt32, ::testing::Combine(
+    } fvalue;
+
+    fvalue.i = x;
+    return fvalue.f;
```
@Leonardo2718 By using `memcpy`, do you mean code like this?
```
float ibits2f(int32_t x) {
    float f;
    memcpy(&f, &x, sizeof(x));
    return f;
}
```

knn-k replied:
FYI.  fvtest/compilertest/tests/X86OpCodesTest.cpp has similar code using `union`.
```
      union
         {
         float    f;
         uint32_t ui32;
         } gold, result;
      gold.f = sqrt(floatDataArray[i]);
      result.f = _fSqrt(floatDataArray[i]);
      OMR_CT_EXPECT_EQ(_fSqrt, gold.ui32, result.ui32);
```

Leonardo2718 replied:
Yes, that's what I meant for `memcpy`.

I'm not terribly worried about the use of a union in the test compiler because the plan is for the Tril tests to replace those tests. :slightly_smiling_face:

knn-k commented:
- Replaced the code using `union` by the code using `memcpy`.
- Added a filter for `fbits2i` and `dbits2l`.

knn-k commented:
Resolved merge conflicts.

knn-k commented:
Resolved merge conflicts.

knn-k commented:
Resolved merge conflicts.

fjeremic commented:
Do we plan on making further progress here? If not we should close the PR.

knn-k commented:
I believe these tests are still useful, and I resolved merge conflicts.
AArch64 passes the new tests.

knn-k commented:
I didn't notice there were some failures on x86-64 Windows and macOS.

Windows:
```
2021-02-05T05:26:13.5923454Z 28: Compilation failed unexpectedly
2021-02-05T05:26:13.5924434Z 28: Input trees: (method return=Int32  (block    (ireturn      (fbits2i        (fconst -nan(ind)) ) ) ) )
2021-02-05T05:26:13.5926508Z 28: [  FAILED  ] TypeConversionTest/FloatBitsToInt32.UsingConst/11, where GetParam() = (-nan(ind), ("fbits2i", 00007FF629607FB0)) (3 ms)

2021-02-05T05:26:13.7928745Z 28: Compilation failed unexpectedly
2021-02-05T05:26:13.7930301Z 28: Input trees: (method return=Int64  (block    (lreturn      (dbits2l        (dconst -nan(ind)) ) ) ) )
2021-02-05T05:26:13.7932001Z 28: [  FAILED  ] TypeConversionTest/DoubleBitsToInt64.UsingConst/11, where GetParam() = (-nan(ind), ("dbits2l", 00007FF629609E80)) (3 ms)
```

macOS:
```
2021-02-05T05:08:53.3978960Z 45: [----------] 38 tests from TypeConversionTest/FloatBitsToInt32
2021-02-05T05:08:53.3979710Z 45: /Users/runner/work/1/s/fvtest/compilertriltest/TypeConversionTest.cpp:1707: Failure
2021-02-05T05:08:53.3980340Z 45:       Expected: exp
2021-02-05T05:08:53.3981190Z 45:       Which is: -4194304
2021-02-05T05:08:53.3981750Z 45: To be equal to: act
2021-02-05T05:08:53.3982270Z 45:       Which is: 2143289344
2021-02-05T05:08:53.3983610Z 45: [  FAILED  ] TypeConversionTest/FloatBitsToInt32.UsingConst/11, where GetParam() = (nan, ("fbits2i", 0x8834ae0)) (1 ms)

2021-02-05T05:08:53.3989090Z 45: [----------] 40 tests from TypeConversionTest/DoubleBitsToInt64
2021-02-05T05:08:53.3989830Z 45: /Users/runner/work/1/s/fvtest/compilertriltest/TypeConversionTest.cpp:1838: Failure
2021-02-05T05:08:53.3990470Z 45:       Expected: exp
2021-02-05T05:08:53.3991340Z 45:       Which is: -2251799813685248
2021-02-05T05:08:53.3991910Z 45: To be equal to: act
2021-02-05T05:08:53.3992640Z 45:       Which is: 9221120237041090560
2021-02-05T05:08:53.3993790Z 45: [  FAILED  ] TypeConversionTest/DoubleBitsToInt64.UsingConst/11, where GetParam() = (nan, ("dbits2l", 0x8837a30)) (1 ms)
```

fjeremic commented:
@genie-omr build all

fjeremic commented:
@knn-k see Power failures on AIX:

```
10:47:47  "/home/jenkins/workspace/Build/fvtest/compilertriltest/TypeConversionTest.cpp", line 1636.35: 1540-0063 (S) The text "{" is unexpected.
10:47:47  "/home/jenkins/workspace/Build/fvtest/compilertriltest/TypeConversionTest.cpp", line 1658.35: 1540-0063 (S) The text "{" is unexpected.
10:47:47  "/home/jenkins/workspace/Build/fvtest/compilertriltest/TypeConversionTest.cpp", line 1700.35: 1540-0063 (S) The text "{" is unexpected.
10:47:47  "/home/jenkins/workspace/Build/fvtest/compilertriltest/TypeConversionTest.cpp", line 1723.35: 1540-0063 (S) The text "{" is unexpected.
```

It looks like RISC-V has problems too with the evaluator. We'll likely need to disable the test there and open up an issue to track fixing it. Similarly for x86 we can just open up an issue and disable the test.

It does pass on all the other platforms so as long as we open up issues and disable the tests properly I'm happy to give my blessing for this to go in. Thanks for refreshing this work.

knn-k commented:
@genie-omr build all

knn-k commented:
@genie-omr build all

knn-k commented:
@genie-omr build all

knn-k commented:
@genie-omr build all

knn-k commented:
I opened Issues #5808 (Windows), #5809 (macOS), and #5810 (RISC-V), and added `SKIP_ON_` lines for them.
I also fixed the compilation errors on AIX and other platforms.

fjeremic commented:
@Leonardo2718 as the assigned committer do you want to shepherd this one?

Leonardo2718 reviewed (approved):
LGTM :+1:
